src/main/java/com/aws/greengrass/mqttclient/MqttClient.java

```diff
@@ -442,37 +485,89 @@ public synchronized void unsubscribe(UnsubscribeRequest request)
+        String prefixOfReservedTopic = "^\\$aws/rules/\\S+?/";
+        if (Pattern.matches(reservedTopicTemplate, topic.toLowerCase())) {
+            // remove the prefix of "$aws/rules/rule-name/"
+            topic = topic.toLowerCase().split(prefixOfReservedTopic, 2)[1];
+        }
+
+        // Topic should not have no more than maximum number of forward slashes (/)
+        if (topic.chars().filter(num -> num == '/').count() > MAX_NUMBER_OF_FORWARD_SLASHES) {
+            String errMsg = String.format("The topic of publish request must have no"
+                    + "more than %d forward slashes (/). This excludes the first 3 slashes in the mandatory segments "
+                    + "for Basic Ingest topics ($AWS/rules/rule-name/).", MAX_NUMBER_OF_FORWARD_SLASHES);
+            throw new SpoolerStoreException(errMsg);
+        }
+
+        // Check the topic size
+        if (topic.getBytes(StandardCharsets.UTF_8).length > MAX_LENGTH_OF_TOPIC) {
+            String errMsg = String.format("The topic size of publish request must be no "
+                            + "larger than {} bytes of UTF-8 encoded characters. This excludes the first "
+                            + "3 mandatory segments for Basic Ingest topics ($AWS/rules/rule-name/).",
+                    MAX_LENGTH_OF_TOPIC);
+            throw new SpoolerStoreException(errMsg);
+        }
     }
 
     @SuppressWarnings({"PMD.AvoidCatchingThrowable", "PMD.PreserveStackTrace"})
     protected CompletableFuture<Integer> publishSingleSpoolerMessage() throws InterruptedException {
         long id = -1L;
         try {
             id = spool.popId();
-            PublishRequest request = spool.getMessageById(id);
+            SpoolMessage spooledMessage = spool.getMessageById(id);
+            PublishRequest request = spooledMessage.getRequest();
             MqttMessage m = new MqttMessage(request.getTopic(), request.getPayload());
 
             long finalId = id;
             return getConnection(false).publish(m, request.getQos(), request.isRetain())
                     .whenComplete((packetId, throwable) -> {
                         // packetId is the SDK assigned ID. Ignore this and instead use the spooler ID
                         if (throwable == null) {
+                            spooledMessage.getFuture().complete(packetId);
                             spool.removeMessageById(finalId);
                             logger.atDebug().kv("id", finalId).kv("topic", request.getTopic())
                                     .log("Successfully published message");
                         } else {
-                            spool.addId(finalId);
-                            logger.atError().log("Failed to publish the message via Spooler and will retry",
-                                    throwable);
+                            if (spooledMessage.getRetried().getAndIncrement()
+                                    < DEFAULT_PUBLISH_RETRIED_COUNTER) {
```

Comment:
I'm not sure how best to handle this. I see the desire to have a way to recover if our message validation isn't good enough, but this also opens up the door to drop QoS 1 messages.

IMO we cannot drop QoS 1 messages. The customer is ultimately in control of the data that they send, and if they have a poison pill, they're doing it to themselves. I don't think we should drop data here, personally.

Reply:
If the Spooler queue is full, we would reject the insertion and leave the Error the message for the customer. Just wondering may be it is fine to leave some Error message and leave the space for the customer to handle this.

Reply:
The challenge is, there is only one queue in the spooler and all the components share the queue. If the queue is blocked, it impacts the whole Greengrass core.